src/main/kotlin/org/openmbee/mms5/Model.kt

```diff
@@ -284,18 +287,29 @@ suspend fun MmsL1Context.queryModel(inputQueryString: String, refIri: String, co
 
         // overwrite the query pattern with a group block
         queryPattern = ElementGroup().apply {
-            // create union between auth failure and user query block
+            // create union between auth failure and user query block as a disjunction
             addElement(ElementUnion().apply {
-                // prevent any bindings to outer scope
-                addElement(ElementFilter(E_Exists(ElementGroup().apply {
-                    // negate authorization query; it must fail in order to produce a binding
-                    addElement(ElementFilter(E_NotExists(ElementGroup().apply {
-                        addElement(patternPlaceholder)
-                    })))
-
-                    // throw error
-                    addElement(ElementService("urn:mms:throw", ElementTriplesBlock()))
-                })))
+                // the first union member matches iff auth fails
+                addElement((ElementGroup().apply {
+                    // create service iri binding in same bgp as negation
+                    addElement(ElementGroup().apply {
+                        // negate authorization query; it must fail in order to produce a binding
+                        addElement(ElementFilter(E_NotExists(ElementGroup().apply {
+                            addElement(patternPlaceholder)
+                        })))
+
+                        // bind(<urn:mms:throw> as ?__mms_service_iri)
+                        addElement(ElementBind(
+                            Var.alloc( "__mms_service_iri"),
+                            NodeValueNode(NodeFactory.createURI("urn:mms:throw"))
+                        ))
+                    })
+
+                    // throw error using ?__mms_service_iri
+                    addElement(ElementService(
+                        NodeFactory.createVariable("__mms_service_iri"),
+                        ElementTriplesBlock(), false))
+                }))
```

Comment:
it's explained in the comments, but the query is carefully designed to lead to the SERVICE block being evaluated when the conditions are not met in order to produce a non-200 response from the SPARQL engine. the filter blocks are disjunctive, so the other one asserts the conditions are met before producing any bindings

Reply:
I think i'm more asking why the filter exists block is needed? Since they're basically mutually exclusive, if the SERVICE block gets evaluated, it already means it'll return a non-200 regardless of whether the other block bind anything (unless I'm misunderstanding how the engine works?) Or is it more of an engine optimization so it doesn't need to evaluate the user query block unless the conditions are met if it's doing the blocks in parallel?

Reply:
yeah its about performance and is essentially there to prevent the engine from evaluating the user query if the auth fails

Reply:
otherwise, a complex user query would end up eating resources while the engine evaluates the 2nd union in parallel (depending on the implementation obviously but this is the case with jena and neptune)